### PR TITLE
No more unsafe `List.head` in SWRA. Closes #1615

### LINKED ISSUE #1615: Error about trying to head an empty list
- develop branch from 0.22
- local backend
- single workflow

I think one of the tasks failed, but got the message below.  cromwell did not exit and I believe it should have.

```
[2016-10-24 14:44:19,47] [error] head of empty list
java.util.NoSuchElementException: head of empty list
        at scala.collection.immutable.Nil$.head(List.scala:420)
        at scala.collection.immutable.Nil$.head(List.scala:417)
        at cromwell.engine.workflow.SingleWorkflowRunnerActor$$anonfun$8.apply(SingleWorkflowRunnerActor.scala:133)
        at cromwell.engine.workflow.SingleWorkflowRunnerActor$$anonfun$8.apply(SingleWorkflowRunnerActor.scala:133)
        at scala.Option.getOrElse(Option.scala:121)
        at cromwell.engine.workflow.SingleWorkflowRunnerActor.cromwell$engine$workflow$SingleWorkflowRunnerActor$$issueReply(SingleWorkflowRunnerActor.scala:133)
        at cromwell.engine.workflow.SingleWorkflowRunnerActor$$anonfun$4.applyOrElse(SingleWorkflowRunnerActor.scala:88)
        at cromwell.engine.workflow.SingleWorkflowRunnerActor$$anonfun$4.applyOrElse(SingleWorkflowRunnerActor.scala:85)
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
        at akka.actor.FSM$class.processEvent(FSM.scala:663)
        at cromwell.engine.workflow.SingleWorkflowRunnerActor.akka$actor$LoggingFSM$$super$processEvent(SingleWorkflowRunnerActor.scala:34)
        at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
        at cromwell.engine.workflow.SingleWorkflowRunnerActor.processEvent(SingleWorkflowRunnerActor.scala:34)
        at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
        at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
        at cromwell.server.CromwellRootActor.aroundReceive(CromwellRootActor.scala:27)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
        at akka.actor.ActorCell.invoke(ActorCell.scala:495)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

```

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 1]
```diff
@@ -32,60 +32,63 @@ import scala.util.{Failure, Try}
  * print out the outputs when complete and reply with a result.
  */
 class SingleWorkflowRunnerActor(source: WorkflowSourceFiles, metadataOutputPath: Option[Path])
-  extends CromwellRootActor with LoggingFSM[RunnerState, RunnerData] {
+  extends CromwellRootActor with LoggingFSM[RunnerState, SwraData] {
 
   import SingleWorkflowRunnerActor._
-  private val backoff = SimpleExponentialBackoff(1 second, 1 minute, 1.2)
+  private val backoff = SimpleExponentialBackoff(1 second, 1 minute, 1.05)
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: I don't have any objection but curious what caused this change

File: engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 2]
```diff
@@ -32,60 +32,63 @@ import scala.util.{Failure, Try}
...
-      val updatedData = data.copy(terminalState = Option(WorkflowFailed)).addFailure(s"Workflow ${data.id.get} transitioned to state Failed")
-      // If there's an output path specified then request metadata, otherwise issue a reply to the original sender.
-      if (metadataOutputPath.isDefined) requestMetadata else issueReply(updatedData)
+        s"CompleteRequest-Workflow-$id-request-${UUID.randomUUID()}")
+      metadataBuilder ! WorkflowOutputs(id)
+      log.info(s"$Tag workflow finished with status '$WorkflowSucceeded'.")
+      goto(RequestingOutputs) using SucceededSwraData(replyTo, id)
+    case Event(RequestComplete((StatusCodes.OK, jsObject: JsObject)), RunningSwraData(replyTo, id)) if jsObject.state == WorkflowFailed =>
+      log.info(s"$Tag workflow finished with status '$WorkflowFailed'.")
+      requestMetadataOrIssueReply(FailedSwraData(replyTo, id, new RuntimeException(s"Workflow $id transitioned to state Failed")))
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: why not the fancy `$WorkflowFailed` thing here?

Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: why not the fancy `$WorkflowFailed` thing here?

File: engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 3]
```diff
@@ -94,22 +97,24 @@ class SingleWorkflowRunnerActor(source: WorkflowSourceFiles, metadataOutputPath:
...
       stay()
   }
 
-  private def requestMetadata: State = {
-    val metadataBuilder = context.actorOf(MetadataBuilderActor.props(serviceRegistryActor), s"MetadataRequest-Workflow-${stateData.id.get}")
-    metadataBuilder ! GetSingleWorkflowMetadataAction(stateData.id.get, None, None)
-    goto (RequestingMetadata)
+  private def requestMetadataOrIssueReply(newData: CompletedSwraData) = if (metadataOutputPath.isDefined) requestMetadata(newData) else issueReply(newData)
+
+  private def requestMetadata(newData: CompletedSwraData) = {
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: perhaps `TerminalSwraData` would be more consistent with our nomenclature?

File: engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 4]
```diff
@@ -118,27 +123,49 @@ class SingleWorkflowRunnerActor(source: WorkflowSourceFiles, metadataOutputPath:
...
+      case f: FailedSwraData => issueFailureReply(f.replyTo, f.failure)
+      case a: AbortedSwraData => issueSuccessReply(a.replyTo)
+
+    }
   }
 
-  private def failAndFinish(e: Throwable): State = {
+  private def failAndFinish(e: Throwable, data: SwraData): State = {
     log.error(e, s"$Tag received Failure message: ${e.getMessage}")
-    issueReply(stateData.addFailure(e))
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: It looks like the old structure had the ability to store multiple failures, though I'm not sure the code was actually structured to be able to capture more than one.  Have we actually lost that ability with these changes?

File: engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala [hunk 5]
```diff
@@ -118,27 +123,49 @@ class SingleWorkflowRunnerActor(source: WorkflowSourceFiles, metadataOutputPath:
...
+    data match {
+      case EmptySwraData =>
+        log.error(e, s"Cannot issue response for {}. Need a 'replyTo' address to issue the response!", e.getMessage)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: not sure why that's an `s` string

Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: also isn't interpolating the `e.getMessage` into the message going to be redundant with what happens passing the exception as the first parameter?

**Comment by mcovarr** [comment]
:+1:

File: engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala (hunk 4 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala]: Yes. Nothing to stop up adding a NEL in the future if we need more than one but I couldn't find an instance where it was actually using more than one entry in the list

**Comment by Horneth** [comment]
👍

**Comment by cjllanwarne** [comment]
Safe to merge - the only centaur fails are due to changed message syntax in centaur and unrelated to this change.

**Pull request merged**